### PR TITLE
Added Consumer driven contract test between bulk scan processer and Sample App

### PR DESCRIPTION
**Pull request by vakepati**
**Does this PR introduce a breaking change?** (check one with "x")
[ ] Yes
[ X] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -17,6 +18,16 @@ version = '0.0.1'
 sourceCompatibility = 1.8
 targetCompatibility = 1.8
 
+
+project.ext {
+  pactVersion = getCheckedOutGitCommitHash()
```
Comment (rkondratowicz) [inline build.gradle]: why change contract version on every new merge?

File: build.gradle [hunk 2]
```diff
@@ -73,7 +93,7 @@ task integration(type: Test, description: 'Runs the integration tests.', group:
   classpath = sourceSets.integrationTest.runtimeClasspath
 
   // set your environment variables here
-   environment("APPINSIGHTS_INSTRUMENTATIONKEY", "test-key")
+  environment("APPINSIGHTS_INSTRUMENTATIONKEY", "test-key")
```
Comment (rkondratowicz) [inline build.gradle]: can be reverted

File: build.gradle [hunk 3]
```diff
@@ -92,6 +112,22 @@ task apiGateway(type: Test) {
...
+  description = "Runs Contract Tests"
+  testClassesDirs = sourceSets.testContract.output.classesDirs
+  classpath = sourceSets.testContract.runtimeClasspath
+  useJUnitPlatform()
+}
+
+pact {
+  publish {
+    pactDirectory = 'target/pacts'
+    pactBrokerUrl = System.getenv("PACT_BROKER_FULL_URL") ?:'http://localhost:80'
```
Comment (rkondratowicz) [inline build.gradle]: missing space

Comment (rkondratowicz) [reply build.gradle]: looks like this env var is never set in tf

File: build.gradle [hunk 4]
```diff
@@ -92,6 +112,22 @@ task apiGateway(type: Test) {
...
+    tags = [System.getenv("PACT_BRANCH_NAME") ?:'ConsumerDrivenContractTest']
```
Comment (rkondratowicz) [inline build.gradle]: same here

File: build.gradle [hunk 5]
```diff
@@ -250,6 +286,16 @@ dependencies {
   testCompile group: 'com.icegreen', name: 'greenmail', version: '1.5.10'
   testCompile group: 'org.apache.commons', name: 'commons-email', version: '1.5'
 
+  testContractCompile sourceSets.main.runtimeClasspath
+  testContractCompile sourceSets.integrationTest.runtimeClasspath
+  testContractCompile  group: 'au.com.dius', name: 'pact-jvm-provider-spring_2.12', version: '3.5.24'
```
Comment (rkondratowicz) [inline build.gradle]: space

File: build.gradle [hunk 6]
```diff
@@ -250,6 +286,16 @@ dependencies {
...
+  testContractCompile group: 'au.com.dius', name: 'pact-jvm-consumer-junit5_2.12', version: '3.5.24'
+  testContractCompile group: 'au.com.dius', name: 'pact-jvm-consumer-java8_2.12', version: '3.5.24'
+  testContractCompile("org.junit.jupiter:junit-jupiter-api:5.3.2")
+  testContractRuntime("org.junit.jupiter:junit-jupiter-engine:5.3.2")
+  testContractImplementation('org.junit.jupiter:junit-jupiter-api:5.3.2')
```
Comment (rkondratowicz) [inline build.gradle]: let's keep it consistent

File: gradle/wrapper/gradle-wrapper.properties [hunk 7]
```diff
@@ -1,5 +1,6 @@
+#Mon Jul 29 10:52:18 BST 2019
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.4.1-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.4.1-all.zip
```
Comment (rkondratowicz) [inline gradle/wrapper/gradle-wrapper.properties]: unrelated

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+import static java.util.UUID.randomUUID;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+
+@ExtendWith(PactConsumerTestExt.class)
+@PactTestFor(providerName = "ocr_validation_service", port = "8889")
+@SpringBootTest(properties = {
+    "ocr_validation_service: localhost:8889",
+    "ocr_validation_service.api.url: localhost:8889"})
```
Comment (rkondratowicz) [inline src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: does this spin up processor service? If so, does it need to?

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+@ContextConfiguration(classes = ConsumerApplication.class)
+public class BulkScanProcesserConsumerDetailsTest {
+
+    private String s2sToken = randomUUID().toString();
+    private String bearerS2sToken = "Bearer " + s2sToken;
+
+    @Autowired
+    private OcrValidationClient client;
+
+    @Pact(state = "Consumer POSTS valid ocrdata ", provider = "ocr_validation_service",
```
Comment (rkondratowicz) [inline src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: typo in 'ocrdata'

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            .path("/forms/PERSONAL/validate-ocr")
+            .method("POST")
+            .headers("ServiceAuthorization", bearerS2sToken)
+            .matchHeader("Content-Type", "application/json;charset=UTF-8")
+            .body(createJsonObject("ocr_data.json"))
+            .willRespondWith()
+            .status(200)
+            .matchHeader("Content-Type", "application/json")
+            .body(newJsonBody((o) -> {
+                o.stringValue("status", "SUCCESS");
```
Comment (rkondratowicz) [inline src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: shall we also check that warning as errors are empty?

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            .uponReceiving("a request to POST an valid ocrdata with success response")
+            .path("/forms/PERSONAL/validate-ocr")
+            .method("POST")
+            .headers("ServiceAuthorization", bearerS2sToken)
+            .matchHeader("Content-Type", "application/json;charset=UTF-8")
+            .body(createJsonObject("ocr_data.json"))
+            .willRespondWith()
+            .status(200)
+            .matchHeader("Content-Type", "application/json")
+            .body(newJsonBody((o) -> {
```
Comment (rkondratowicz) [inline src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: what's `o`?
`(` `)` not needed

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+                o.stringValue("status", "SUCCESS");
+            }).build())
+            .toPact();
+    }
+
+    @Test
+    @PactTestFor(pactMethod = "executePostSubmissionOfOcrDataWithSuccessPact")
+    public void verifyExecutePostSubmissionWithSuccessPact() throws Exception {
+        ValidationResponse res = client.validate("http://localhost:8889", getOcrData(), "PERSONAL", s2sToken);
+        assertThat(res.status, equalTo(Status.SUCCESS));
```
Comment (rkondratowicz) [inline src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: we use AssertJ on the project

File: src/integrationTest/resources/zipcontents/invalid_json/metadatajson.json [hunk 13]
```diff
@@ -16,6 +16,6 @@
...
         "file_name": "test1.pdf",
         "notes": ""
-      },
+      }
```
Comment (rkondratowicz) [inline src/integrationTest/resources/zipcontents/invalid_json/metadatajson.json]: ?

File: .gitignore [hunk 14]
```diff
@@ -14,6 +14,7 @@
 ### IntelliJ IDEA ###
 .idea
 /out
+/target
```
Comment (timja) [inline .gitignore]: ? that's a maven folder

File: build.gradle (hunk 1 above)
Comment (vakepati) [reply build.gradle]: It will publish the latest contract for every commit.

File: gradle/wrapper/gradle-wrapper.properties (hunk 7 above)
Comment (vakepati) [reply gradle/wrapper/gradle-wrapper.properties]: removed

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java (hunk 8 above)
Comment (vakepati) [reply src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: it won't spin up the processor service,mocking the ocr validation service.

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java (hunk 10 above)
Comment (vakepati) [reply src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: done +1

Comment (vakepati) [reply src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: done!

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java (hunk 12 above)
Comment (vakepati) [reply src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: for PactTest Annotations we need JUnit Jupiter library.

File: build.gradle (hunk 2 above)
Comment (vakepati) [reply build.gradle]: done!

File: build.gradle (hunk 3 above)
Comment (vakepati) [reply build.gradle]: As of now not yet configured in the pipeline.

File: src/integrationTest/resources/zipcontents/invalid_json/metadatajson.json (hunk 13 above)
Comment (vakepati) [reply src/integrationTest/resources/zipcontents/invalid_json/metadatajson.json]: there was extra comma in metafile,so i removed it.

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java (hunk 12 above)
Comment (rkondratowicz) [reply src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: AssertJ is just for assertions, JUnit 5 can still be used

File: src/integrationTest/resources/zipcontents/invalid_json/metadatajson.json (hunk 13 above)
Comment (rkondratowicz) [reply src/integrationTest/resources/zipcontents/invalid_json/metadatajson.json]: unrelated to your PR. perhaps the comma was there to make the json invalid as the folder names suggests? (`invalid_json`)

File: build.gradle (hunk 1 above)
Comment (rkondratowicz) [reply build.gradle]: imho contract version shouldn't change if the contract doesn't change.

File: gradle/wrapper/gradle-wrapper.properties (hunk 7 above)
Comment (rkondratowicz) [reply gradle/wrapper/gradle-wrapper.properties]: still here

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java (hunk 10 above)
Comment (rkondratowicz) [reply src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: needs some formatting.  also those arrays should be empty I think, not null

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/ConsumerApplication.java [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+@ComponentScan(basePackages = "uk.gov.hmcts.reform.bulkscanprocessor.ocrvalidation.client")
+public class ConsumerApplication {
+    @Bean
+    public RestTemplate restTemplate() {
+        return new RestTemplate();
+    }
+
+
+    @Bean
+    ObjectMapper objectMapper() {
```
Comment (rkondratowicz) [inline src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/ConsumerApplication.java]: missing `public`

File: build.gradle [hunk 16]
```diff
@@ -250,6 +286,14 @@ dependencies {
   testCompile group: 'com.icegreen', name: 'greenmail', version: '1.5.10'
   testCompile group: 'org.apache.commons', name: 'commons-email', version: '1.5'
 
+  testContractCompile sourceSets.main.runtimeClasspath
+  testContractCompile sourceSets.integrationTest.runtimeClasspath
+  testContractCompile group: 'au.com.dius', name: 'pact-jvm-provider-spring_2.12', version: '3.5.24'
+  testContractCompile group: 'au.com.dius', name: 'pact-jvm-consumer-junit5_2.12', version: '3.5.24'
+  testContractCompile group: 'au.com.dius',name: 'pact-jvm-consumer-java8_2.12', version: '3.5.24'
```
Comment (rkondratowicz) [inline build.gradle]: missing space

File: build.gradle [hunk 17]
```diff
@@ -9,6 +9,7 @@ plugins {
   id 'org.owasp.dependencycheck' version '5.2.0'
   id 'com.github.ben-manes.versions' version '0.21.0'
   id 'org.sonarqube' version '2.7.1'
+  id 'au.com.dius.pact' version '3.5.24'
```
Comment (rkondratowicz) [inline build.gradle]: can we use the latest? (`3.6.11`)?

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+
+
+@ExtendWith(PactConsumerTestExt.class)
+@PactTestFor(providerName = "ocr_validation_service", port = "8889")
+@SpringBootTest(properties = {
+    "ocr_validation_service: localhost:8889",
+    "ocr_validation_service.api.url: localhost:8889"})
+@ContextConfiguration(classes = ConsumerApplication.class)
+public class BulkScanProcesserConsumerDetailsTest {
```
Comment (rkondratowicz) [inline src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: typo in 'processor'

Comment (rkondratowicz) [reply src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: would be good if this somehow indicated that it's testing the validation endpoint

File: src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java (hunk 8 above)
Comment (rkondratowicz) [reply src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: what are those 2 properties? what's using it?

Comment (lgonczar) [reply src/consumerDrivenContractTest/java/uk/gov/hmcts/reform/bulkscanprocessor/contract/tests/BulkScanProcesserConsumerDetailsTest.java]: do those tests work?
